hare3/malfeasance.go

```diff
@@ -0,0 +1,99 @@
+package hare3
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/spacemeshos/go-scale"
+	"go.uber.org/zap"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/malfeasance/wire"
+	"github.com/spacemeshos/go-spacemesh/signing"
+	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/sql/atxs"
+)
+
+const (
+	hareEquivocate = "hare_eq"
+)
+
+type MalfeasanceHandler struct {
+	logger *zap.Logger
+	db     sql.Executor
+
+	edVerifier *signing.EdVerifier
+}
+
+type MalfeasanceOpt func(*MalfeasanceHandler)
+
+func WithMalfeasanceLogger(logger *zap.Logger) MalfeasanceOpt {
+	return func(mh *MalfeasanceHandler) {
+		mh.logger = logger
+	}
+}
+
+func NewMalfeasanceHandler(
+	db sql.Executor,
+	edVerifier *signing.EdVerifier,
+	opt ...MalfeasanceOpt,
+) *MalfeasanceHandler {
+	mh := &MalfeasanceHandler{
+		logger: zap.NewNop(),
+		db:     db,
+
+		edVerifier: edVerifier,
+	}
+	for _, o := range opt {
+		o(mh)
+	}
+	return mh
+}
+
+func (mh *MalfeasanceHandler) HandleHareEquivocation(
+	ctx context.Context,
+	data scale.Type,
+) (types.NodeID, []string, error) {
+	var (
+		firstNid types.NodeID
+		firstMsg wire.HareProofMsg
+	)
+	hp, ok := data.(*wire.HareProof)
+	if !ok {
+		return types.EmptyNodeID, []string{hareEquivocate}, errors.New("wrong message type for hare equivocation")
+	}
+	for _, msg := range hp.Messages {
```

Comment:
Ditto: I think the code could be simpler if it compared `hp.Messages[0]` and `hp.Messages[1]`.

Reply:
Updated